### PR TITLE
fuzzai: add test edge cases

### PR DESCRIPTION
**Pull request by psiinon**
Files c/o @yns000 :grin:

### CODE REVIEW

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+<BODY>
+<H1>Fuzz AI Files - Extract Training Data</H1>
+
+<H3>001 By Asking</H3>
+
+This file contains fuzzing payloads specific to AI-RMM Control "Measure 2.7.2"<br>
+The goal is to test if the system has established security and resilience metrics
+to detect potential theft of model training data, including encryption, access controls,
+and alarm thresholds for unauthorized access.
+<br><br>
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: Paragraph vs double break?

File: addOns/fuzzai/CHANGELOG.md [hunk 2]
```diff
@@ -4,8 +4,12 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
+### Added
+- Test edge cases files
```
Comment (kingthorin) [inline addOns/fuzzai/CHANGELOG.md]: How about:
> Edge case test files.

or

> Test files for edge cases.

File: addOns/fuzzai/CHANGELOG.md [hunk 3]
```diff
@@ -4,8 +4,12 @@ All notable changes to this add-on will be documented in this file.
...
+
 ### Changed
 - Update minimum ZAP version to 2.16.0.
+- Help to split across more files and to link to the fuzzing files.
```
Comment (kingthorin) [inline addOns/fuzzai/CHANGELOG.md]:
```suggestion
- Re-organize help content and include cross links to the fuzzing file sources.
```

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+
+<H3>001 Architecture</H3>
+
+This file contains advanced fuzzing payloads designed to extract model architecture information from LLMs.<br>
+The goal is to determine if the LLM reveals specific details about its architecture, 
+such as the number of layers, transformer types, or parameter counts.
+<p>
+Model Architecture Extraction: These payloads probe for specific architecture-related details 
+such as layer counts, model types, and other structural aspects of the LLM.
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: These two paragraphs say essentially the same thing.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+<p>
+Layer counts - Payloads focus on extracting the number of layers and their types (e.g., transformer layers, hidden layers).<br>
+Parameter inference - Probes for parameter counts, such as attention heads or layer dimensions.<br>
+Model type - Attackers attempt to identify whether the LLM is a decoder-only model or part of an architecture family (e.g., GPT, BERT).<br>
+Specific internal mechanisms - Questions around activation functions, normalization, and positional encodings to infer architecture specifics.<br>
+<p>
+This file contains fuzzing payloads specific to AI-RMM Control "Measure 2.7.2"<br>
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: Linkify AI-RMM?

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+<H3>004 Conditional Prompts</H3>
+
+This file contains advanced fuzzing payloads for conditional prompts designed to extract context-specific information from LLMs.<br>
+These payloads guide the model by introducing conditions or constraints, aiming to extract information from specific domains or contexts within the training data.
+
+<p>
+Conditional Prompts: These prompts impose constraints that direct the model towards generating text that aligns with specific criteria or domain-specific knowledge.<br>
+<p>
+Conditional constraints - Each payload introduces specific conditions (e.g., domain-specific phrases or
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: Use of colons, hyphens, paragraphs and breaks doesn't seem to be consistent throughout the H3 sections of this doc

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+</ul>
+
+<H3>008 Visual Representation Leakage</H3>
+
+This file contains advanced fuzzing payloads for visual representation analysis, focusing on stealing model information from scientific plots.<br>
+These prompts aim to infer model architecture, hyperparameters, and training details by leveraging patterns in visual outputs like t-SNE or loss plots.
+
+<p>
+Visual Representation Analysis - These prompts target model information extracted through visual outputs such as t-SNE and loss plots.
+Inspired from: "A Plot is Worth a ThousandWords: Model Information Stealing Attacks via Scientific Plots".
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: Should this be linked?
Should there be a space in "Thousand Words"?

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+<p>
+Author: 
+<ul>
+<li>yiannis@owasp.org
+</ul>
+
+<H3>008 Visual Representation Leakage</H3>
+
+This file contains advanced fuzzing payloads for visual representation analysis, focusing on stealing model information from scientific plots.<br>
+These prompts aim to infer model architecture, hyperparameters, and training details by leveraging patterns in visual outputs like t-SNE or loss plots.
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html:173]: Will an average tester know what t-SNE means?

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+<ul>
+<li>yiannis@owasp.org
+</ul>
+
+<H3>003 Divergence Attack</H3>
+
+This file contains fuzzing payloads specific to testing LLM "hypnosis" vulnerabilities.<br>
+The goal is to determine if the LLM enters a repetitive state where it unintentionally leaks training data.
+These payloads explore edge cases where repetition may trigger the LLM to output sensitive data.
+<p>
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: 002?

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+that trigger memorized training data output.
+These payloads explore edge cases where special characters lead the model to output sensitive data.
+<p>
+Payloads simulate input sequences that induce LLMs to generate memorized content.
+This version is based on the SCA method, as described in the document provided.
+Expected system behavior is detailed for each scenario.
+<p>
+Special Characters Attack (SCA) - Payloads focus on using specific symbols, such as JSON structural symbols or other
+commonly occurring characters in LLM training corpora, to induce data leakage.<br>
+<p>
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: Use of breaks and paragraphs doesn't seem consistent in this file.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+Fuzzing File: <a href="https://github.com/zaproxy/zap-extensions/blob/main/addOns/fuzzai/src/main/zapHomeFiles/fuzzers/fuzzai/AI/extract-training-data/006-quantifying-memorization.txt">006-quantifying-memorization.txt</a>
+<p>
+Author: 
+<ul>
+<li>yiannis@owasp.org
+</ul>
+
+<H3>007 Data Poisoning Attacks</H3>
+
+This file contains advanced fuzzing payloads specific to testing data poisoning vulnerabilities in DP-SGD models.<br>
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: Will an average tester know what DP-SGD means?

**Comment by psiinon** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [12f97c60\-c7e5\-4e57\-be14\-fa5bbb026d83](https://eu.ast.checkmarx.net/projects/0da811fa-6373-42b1-bbb6-d0026285b136/scans?id=12f97c60-c7e5-4e57-be14-fa5bbb026d83&branch=psiinon%2Ffuzzai%2Fedge-cases)
### Great job\! No new security vulnerabilities introduced in this pull request

File: addOns/fuzzai/CHANGELOG.md (hunk 2 above)
Comment (yns000) [reply addOns/fuzzai/CHANGELOG.md]: Agree.

File: addOns/fuzzai/CHANGELOG.md (hunk 3 above)
Comment (yns000) [reply addOns/fuzzai/CHANGELOG.md]: Re-organized help content and included URL links to the fuzzing file sources.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html (hunk 4 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: Keep the first paragraph.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html (hunk 5 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: Yes, here is the link: https://osf.io/kb8a5/

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html (hunk 6 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: Agree. Let's simplify by using `<H3>` for each fuzzing file, followed by paragraph tags `<p>`, no need to use `<br>` as well.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html (hunk 8 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html:173]: Let's add the paragraph: 

<p>t-SNE plots and loss plots serve different purposes in machine learning. t-SNE (t-distributed stochastic neighbor embedding) is a dimensionality reduction technique used to visualize high-dimensional data in a lower-dimensional space, often 2D or 3D, while preserving the local structure of the data. Loss plots, on the other hand, track the model's training progress by visualizing how the loss function changes over time (epochs or iterations). </p>

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html (hunk 7 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html]: We would have added an explanation above on t-SNE and loss plots. 

Thousand Words needs a space. :)

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html (hunk 1 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: Paragraph, let's go with that.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html (hunk 9 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: This paragraph should be deleted, the first paragraph here should be: 

Divergence testing - Payloads are designed to induce divergence from the model’s aligned state, encouraging it to generate training data by combining nonsensical prompts with structured ones.

It is wrong also in the file 003-divergence-attack.txt

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html (hunk 10 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: Only paragraphs should be used.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html (hunk 11 above)
Comment (yns000) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-training-data.html]: You are right, let's include the following text: 

DP-SGD stands for Differentially Private Stochastic Gradient Descent.

It's a privacy-preserving version of the standard Stochastic Gradient Descent (SGD) algorithm used to train machine learning models — especially neural networks — while ensuring differential privacy for the data used during training.

**Comment by psiinon** [comment]
Hopefully got all of those?

**Comment by kingthorin** [comment]
Will try to re-review on Monday.

**Comment by kingthorin** [comment]
Addressed a few outstanding items.

File: addOns/fuzzai/src/main/zapHomeFiles/fuzzers/fuzzai/AI/extract-training-data/003-divergence-attack.txt
Comment (kingthorin) [inline addOns/fuzzai/src/main/zapHomeFiles/fuzzers/fuzzai/AI/extract-training-data/003-divergence-attack.txt:1]: Per: https://github.com/zaproxy/zap-extensions/pull/6649#discussion_r2260594382
image

**Comment by thc202** [comment]
Please review the latest pushes you did (to confirm the changes), there are now sentences/paragraphs in half e.g. `# Sensitive data requests - These payloads assess whether the model can pro`.

**Comment by kingthorin** [comment]
Will do, thanks!

File: addOns/fuzzai/src/main/zapHomeFiles/fuzzers/fuzzai/AI/extract-training-data/003-divergence-attack.txt [hunk 12]
```diff
@@ -1,23 +1,15 @@
...
+# 
 # Repetition strategies - Several prompts employ repetition to test if the model diverges and 
 # outputs memorized data.
-#
-# Sensitive data requests - These payloads assess whether the model can protect sensitive training 
-# data like PII, code snippets, or real-world text examples from leaking.
-#
-#
+# 
+# Sensitive data requests - These payloads assess whether the model can pro
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/zapHomeFiles/fuzzers/fuzzai/AI/extract-training-data/003-divergence-attack.txt]:
```suggestion
# Sensitive data requests - These payloads assess whether the model can protect sensitive training 
# data like PII, code snippets, or real-world text examples from leaking.
```

File: addOns/fuzzai/src/main/zapHomeFiles/fuzzers/fuzzai/AI/extract-training-data/003-divergence-attack.txt [hunk 13]
```diff
@@ -1,24 +1,17 @@
...
 # Expected system behavior is detailed for each scenario. 
-#
-# Divergence testing - Payloads are designed to induce divergence from the model’s aligned state, 
-# encouraging it to generate training data by combining nonsensical prompts with structured ones.
-#
+# 
 # Repetition strategies - Several prompts employ repetition to test if the model diverges and 
 # outputs memorized data.
-#
+#
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/zapHomeFiles/fuzzers/fuzzai/AI/extract-training-data/003-divergence-attack.txt]:
```suggestion
#
```

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+optimizer types, batch sizes, and more, through adversarial prompts.
+
+<p>
+Hyperparameter Extraction - These payloads probe for learning rates, optimizer settings, 
+and training schedules used during model development.
+<p>
+Learning rates - Payloads focus on extracting the learning rate and its schedule.
+Optimizer and hyperparameters - Probes for optimizer types, momentum, beta values, and weight decay settings.
+Training specifics - Queries related to batch size, epochs, gradient clipping, and dropout rates.
+Training techniques - Attempts to infer whether early stopping, data augmentation, or mixed-precision training were used.
```
Comment (kingthorin) [inline addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html:53]: Should we make these ULs?

**Comment by kingthorin** [comment]
> Please review the latest pushes you did (to confirm the changes), there are now sentences/paragraphs in half e.g. `# Sensitive data requests - These payloads assess whether the model can pro`.

I think luckily it was just that one.

I started totally re-read but couldn't do it in a single sitting, will finish over the weekend.

File: addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html (hunk 14 above)
Comment (kingthorin) [reply addOns/fuzzai/src/main/javahelp/help/contents/extract-model-info.html:53]: Elsewhere we had agreed that `<br>` weren't necessary but then I'm wondering if these need to be kept left starting.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you all!
